### PR TITLE
fix(training): lr scheduler doesn't work properly in distributed scenarios

### LINKED ISSUE #8236: Wrong learning rate scheduler training step count for examples with multi-gpu when setting `--num_train_epochs`
### Describe the bug

I think there are still some problems with the learning rate scheduler. This is resolved when you set `--max_train_steps`, as discussed in #3954 , but not completely.

For example, the code snippet https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L816-L833 . I paste it here:
```python
# Scheduler and math around the number of training steps.
overrode_max_train_steps = False
num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
if args.max_train_steps is None:
    args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
    overrode_max_train_steps = True

lr_scheduler = get_scheduler(
    args.lr_scheduler,
    optimizer=optimizer,
    num_warmup_steps=args.lr_warmup_steps * accelerator.num_processes,
    num_training_steps=args.max_train_steps * accelerator.num_processes,
)

# Prepare everything with our `accelerator`.
unet, optimizer, train_dataloader, lr_scheduler = accelerator.prepare(
    unet, optimizer, train_dataloader, lr_scheduler
)
```

When setting `--num_train_epochs` instead of `--max_train_steps`, the calculation of `num_update_steps_per_epoch` is incorrect because `train_dataloader` has not yet been wrapped by `accelerator.prepare`. Consequently, `args.max_train_steps` is roughly `num_processes` times the actual value. This discrepancy leads to unintended values being passed into the `get_scheduler` function.

In fact, the logic here is quite confusing. It seems like a refactoring might be necessary.

### Reproduction

```diff
accelerate launch --mixed_precision="fp16" train_text_to_image.py \
  ...
-  --max_train_steps=15000 \
+  --num_train_epochs=100 \
  ...
```

### Logs

_No response_

### System Info

- `diffusers` version: 0.27.2
- Platform: macOS-10.16-x86_64-i386-64bit
- Python version: 3.9.17
- PyTorch version (GPU?): 2.0.1 (False)
- Huggingface_hub version: 0.20.3
- Transformers version: 4.30.0
- Accelerate version: 0.21.0
- xFormers version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@sayakpaul @yiyixuxu @eliphatfs

### PR DESCRIPTION
**Pull request by geniuspatrick**
# What does this PR do?

### TL;DR

In a distributed training scenario, passing the argument `--num_train_epochs` to any of the training scripts disrupts the functioning of the learning rate scheduler. Essentially, the learning rate decays `num_processes` times slower than expected. Related issues #8236, #3954, and PR #3983 shed further light on this.

### Explanation

In our training setup, we utilize `accelerator` instead of PyTorch's native [DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler) when creating the `train_dataloader`. This means we create the `train_dataloader` directly as if for standalone training and subsequently employ `accelerator.prepare` to shard the samples across different processes.

When referencing `step` in training scripts such as `lr_warmup_steps`, `max_train_steps`, etc., we're indicating the  optimizing step. In essence, each step consumes `num_processes * gradient_accumulation_steps` batches of data. In the script, the learning rate scheduler is initialized before `accelerator.prepare` is called. At this stage, the `train_dataloader` hasn't yet sharded the samples, specifically the batched samples.

To accurately calculate `num_update_steps_per_epoch`, we need the length of the `train_dataloader` after distributed sharding. How do we achieve this? Typically, `accelerator.prepare` replaces `train_dataloader.batch_sampler` with [BatchSamplerShard](https://github.com/huggingface/accelerate/blob/b24a0ef5dbb4e0408074f92f84683608eab474af/src/accelerate/data_loader.py#L100). The length of the distributed sharded `train_dataloader` (still a [DataLoader](https://github.com/pytorch/pytorch/blob/92bc444ee32f578ee9ebc2ecb76b283881f806a8/torch/utils/data/dataloader.py#L124) instance) becomes the length of `BatchSamplerShard`. Hence, we derive a formula for estimating the length of the sharded `train_dataloader`, which aligns with current training scripts (where `accelerator.prepare` is called with no extra arguments).

As per `accelerator` principles, the **_prepared_** scheduler calls the `step()` of the **_unprepared_** scheduler `num_processes` times at each optimizing step (once gradient accumulation is completed). This necessitates dividing `num_*_steps_for_scheduler` by `gradient_accumulation_steps` and multiplying it by `num_processes`.

Feeling a bit confused? Not to worry, let's visualize it.

### Experiments

We utilize [Fine-tuning for text2image with LoRA](https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image_lora.py) as an example. Below is the training command:

```bash
export MODEL_NAME="CompVis/stable-diffusion-v1-4"
export DATASET_NAME="lambdalabs/naruto-blip-captions"

# Example of --num_train_epochs
accelerate launch examples/text_to_image/train_text_to_image_lora.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --dataset_name=$DATASET_NAME \
  --resolution=128 --random_flip --max_train_samples=171 \
  --train_batch_size=4 \
  --num_train_epochs=6 \
  --learning_rate=1e-04 --lr_scheduler="cosine_with_restarts" --lr_warmup_steps=3 \
  --gradient_accumulation_steps=5 \
  --seed=42 \
  --output_dir="sd-pokemon-model-lora-epoch"

# Example of --max_train_steps
accelerate launch examples/text_to_image/train_text_to_image_lora.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --dataset_name=$DATASET_NAME \
  --resolution=128 --random_flip --max_train_samples=171 \
  --train_batch_size=4 \
  --max_train_steps=30 \
  --learning_rate=1e-04 --lr_scheduler="cosine_with_restarts" --lr_warmup_steps=3 \
  --gradient_accumulation_steps=5 \
  --seed=42 \
  --output_dir="sd-pokemon-model-lora-step"
```

The hyper-parameters are:

- Batch size: 4
- Number of processes (n_gpus): 2
- Dataset length: 171
- Gradient accumulation steps: 5

Thus,
```python
len_dataloader_standalone = ceil(171/4) = 43
len_dataloader_distribute = ceil(43/2) = 22
num_update_steps_per_epoch = ceil(22/5) = 5
```
And `epochs=6` is equivalent to `steps=30`.

Additionally, introducing the argument `num_cycles=2` to the function `get_scheduler` exacerbates the error.

#### Before the PR
`--num_train_epochs`
截屏2024-05-29 18 32 08

`--max_train_steps`
截屏2024-05-29 18 34 14

#### After the PR
`--num_train_epochs`
截屏2024-05-29 18 34 26

`--max_train_steps`
截屏2024-05-29 18 34 40

Fixes #8236

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sayakpaul @eliphatfs

### CODE REVIEW

File: examples/text_to_image/train_text_to_image_lora.py [hunk 1]
```diff
@@ -657,17 +657,20 @@ def collate_fn(examples):
     )
 
     # Scheduler and math around the number of training steps.
-    overrode_max_train_steps = False
-    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
+    # Check pr xxx for detailed explanation.
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:660]: Feel free to replace the "xxx" :-)

File: examples/text_to_image/train_text_to_image_lora.py [hunk 2]
```diff
@@ -677,8 +680,14 @@ def collate_fn(examples):
...
     # We need to recalculate our total training steps as the size of the training dataloader may have changed.
     num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
-    if overrode_max_train_steps:
+    if args.max_train_steps is None:
         args.max_train_steps = args.num_train_epochs * num_update_steps_per_epoch
+        if num_training_steps_for_scheduler != args.max_train_steps * accelerator.num_processes:
+            logger.warning(
+                "The length of the 'train_dataloader' after 'accelerator.prepare' does not match "
+                "the length that was expected when the learning rate scheduler was created. "
+                "This inconsistency may result in the learning rate scheduler not functioning properly."
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_lora.py:689]: Should we also include the values or "The length of the 'train_dataloader'" and "the length that was expected when the learning rate scheduler was created"?

**Review by sayakpaul — APPROVED** [review]
I just have minor comments but this is very very nicely done. Thanks so much!

File: examples/text_to_image/train_text_to_image_lora.py (hunk 2 above)
Comment (geniuspatrick) [reply examples/text_to_image/train_text_to_image_lora.py]: > Should we also include the values or "The length of the 'train_dataloader'" and "the length that was expected when the learning rate scheduler was created"?

Yep, the values are included!

**Comment by sayakpaul** [comment]
Sorry for pinging late but @geniuspatrick could we keep the changes in this PR to a bare minimum i.e., targeting a single script only and then opening the rest to the community? That will be easier to manage IMO.

File: examples/text_to_image/train_text_to_image_lora.py (hunk 2 above)
Comment (geniuspatrick) [reply examples/text_to_image/train_text_to_image_lora.py]: > Do we have drop_last settings or similar that may cause this to happen?

For all current period training scripts, the answer is no. Our estimate of the length of the sliced dataloader is always correct, and this warning message is never triggered.

**Comment by geniuspatrick** [comment]
> Sorry for pinging late but @geniuspatrick could we keep the changes in this PR to a bare minimum i.e., targeting a single script only and then opening the rest to the community? That will be easier to manage IMO.

OK, I'll change the script `examples/text_to_image/train_text_to_image_lora.py` only.

**Comment by geniuspatrick** [comment]
Hi, @sayakpaul . I think it's ready now. Any suggestions?

**Review by sayakpaul — APPROVED** [review]
Thanks a ton!

**Comment by geniuspatrick** [comment]
Hi,  @sayakpaul  here's a TODO list for follow-up contributions from the community.

### What should be changed

- advanced_diffusion_training
    - [ ] train_dreambooth_lora_sd15_advanced.py
    - [ ] train_dreambooth_lora_sdxl_advanced.py
- consistency_distillation
    - [ ] train_lcm_distill_lora_sdxl.py
- controlnet
    - [ ] train_controlnet.py
    - [ ] train_controlnet_sdxl.py
- custom_diffusion
    - [ ] train_custom_diffusion.py
- dreambooth
    - [ ] train_dreambooth.py
    - [ ] train_dreambooth_lora.py
    - [ ] train_dreambooth_lora_sdxl.py
- instruct_pix2pix
    - [ ] train_instruct_pix2pix.py
    - [ ] rain_instruct_pix2pix_sdxl.py
- kandinsky2_2/text_to_image
    - [ ] train_text_to_image_decoder.py
    - [ ] train_text_to_image_prior.py
    - [ ] train_text_to_image_lora_decoder.py
    - [ ] train_text_to_image_lora_prior.py
- research_projects
    - [ ] consistency_training/train_cm_ct_unconditional.py
    - [ ] diffusion_dpo/train_diffusion_dpo.py
    - [ ] diffusion_dpo/train_diffusion_dpo_sdxl.py
    - [ ] diffusion_orpo/train_diffusion_orpo_sdxl_lora.py
    - [ ] dreambooth_inpaint/train_dreambooth_inpaint.py
    - [ ] dreambooth_inpaint/train_dreambooth_inpaint_lora.py
    - [ ] instructpix2pix_lora/train_instruct_pix2pix_lora.py
    - [ ] intel_opts/textual_inversion/textual_inversion_bf16.py
    - [ ] intel_opts/textual_inversion_dfq/textual_inversion.py
    - [ ] lora/train_text_to_image_lora.py
    - [ ] multi_subject_dreambooth/train_multi_subject_dreambooth.py
    - [ ] multi_token_textual_inversion/textual_inversion.py
    - [ ] onnxruntime/text_to_image/train_text_to_image.py
    - [ ] onnxruntime/textual_inversion/textual_inversion.py
    - [ ] onnxruntime/unconditional_image_generation/train_unconditional.py
    - [ ] realfill/train_realfill.py
    - [ ] scheduled_huber_loss_training/dreambooth/train_dreambooth.py
    - [ ] scheduled_huber_loss_training/dreambooth/train_dreambooth_lora.py
    - [ ] scheduled_huber_loss_training/dreambooth/train_dreambooth_lora_sdxl.py
    - [ ] scheduled_huber_loss_training/text_to_image/train_text_to_image.py
    - [ ] scheduled_huber_loss_training/text_to_image/train_text_to_image_sdxl.py
    - [ ] scheduled_huber_loss_training/text_to_image/train_text_to_image_lora.py
    - [ ] scheduled_huber_loss_training/text_to_image/train_text_to_image_lora_sdxl.py
- t2i_adapter
    - [ ] train_t2i_adapter_sdxl.py
- text_to_image
    - [ ] train_text_to_image.py
    - [ ] train_text_to_image_sdxl.py
    - [x] train_text_to_image_lora.py
    - [ ] train_text_to_image_lora_sdxl.py
- textual_inversion
    - [ ] textual_inversion.py
    - [ ] textual_inversion_sdxl.py
- unconditional_image_generation
    - [ ] train_unconditional.py
- wuerstchen
    - [ ] text_to_image/train_text_to_image_prior.py
    - [ ] text_to_image/train_text_to_image_lora_prior.py

### What should **NOT** be changed

#### Category 1

The script does not have the argument --num_train_epochs.

- amused
    - train_amused.py
- research_projects
    - multi_subject_dreambooth_inpainting/train_multi_subject_dreambooth_inpainting.py

#### Category 2

Distributed dataset sharding is done by WebDataset, not accelerator.

- consistency_distillation
    - train_lcm_distill_sd_wds.py
    - train_lcm_distill_sdxl_wds.py
    - train_lcm_distill_lora_sd_wds.py
    - train_lcm_distill_lora_sdxl_wds.py
- research_projects
    - controlnet/train_controlnet_webdataset.py
    - diffusion_orpo/train_diffusion_orpo_sdxl_lora_wds.py

BTW, if you need more extra hands, I would like to help!

**Comment by sayakpaul** [comment]
Great! Thank you so much, @geniuspatrick!

I will create an issue similar to https://github.com/huggingface/diffusers/issues/6545 so that the community can easily pick them up.

**Comment by AbraarArique** [comment]
Thank you @sayakpaul and @geniuspatrick for fixing this, much appreciated!

But I have a quick question: why is `lr_warmup_steps` multiplied by `num_processes`?

```python
# Line 1092
num_warmup_steps_for_scheduler = args.lr_warmup_steps * accelerator.num_processes
```

For example, if my Dataloader length (e.g. steps per epoch) is 64, and for simplicity let's say I want to warm up to 32 steps.

With 1 epoch and gradient accumulation steps = 1, the current code works correctly for 1 GPU:

```python
num_warmup_steps_for_scheduler = args.lr_warmup_steps * accelerator.num_processes # 32 * 1 = 32
len_train_dataloader_after_sharding = math.ceil(len(train_dataloader) / accelerator.num_processes) # 64 / 1 = 64
num_update_steps_per_epoch = math.ceil(len_train_dataloader_after_sharding / args.gradient_accumulation_steps) # 64 / 1 = 64
num_training_steps_for_scheduler = (
    args.num_train_epochs * num_update_steps_per_epoch * accelerator.num_processes # 1 * 64 * 1 = 64
)
```

But with 2 GPUs, the `num_training_steps_for_scheduler` stays the same (64), but `num_warmup_steps_for_scheduler` doubles to 64.

```python
num_warmup_steps_for_scheduler = args.lr_warmup_steps * accelerator.num_processes # 32 * 2 = 64
len_train_dataloader_after_sharding = math.ceil(len(train_dataloader) / accelerator.num_processes) # 64 / 2 = 32
num_update_steps_per_epoch = math.ceil(len_train_dataloader_after_sharding / args.gradient_accumulation_steps) # 32 / 1 = 32
num_training_steps_for_scheduler = (
    args.num_train_epochs * num_update_steps_per_epoch * accelerator.num_processes # 1 * 32 * 2 = 64
)
```

Shouldn't `num_warmup_steps_for_scheduler` still be 32? Sorry if I'm missing something...

**Comment by eliphatfs** [comment]
In each gradient step, the lr scheduler is advanced for num_processes steps in accelerator, if my memory serves me right. This is counterintuitive.

**Comment by AbraarArique** [comment]
> In each gradient step, the lr scheduler is advanced for num_processes steps in accelerator, if my memory serves me right. This is counterintuitive.

@eliphatfs Right, but to clarify, the `num_training_steps_for_scheduler` doesn't change much with the number of processes/GPUs, does it?

Regardless of 1, 2, or more GPUs, the `num_training_steps_for_scheduler` is still `64` in my example above.

So assuming I want to warm up for half of all steps, the `num_warmup_steps_for_scheduler` should be `32` regardless of the number of processes.

But the current code increases/scales the `num_warmup_steps_for_scheduler` based on `num_processes`, but the total `num_training_steps_for_scheduler` doesn't change. Shouldn't that break things?

**Comment by bghira** [comment]
this is actually incorrect...

image

The disconnected line is with the `* accelerator.num_processes` on the warmup steps resuming at 1000 steps with 3 GPUs.

The learning rate is resumed at the point where it would have been at T=333.

Removing the multiplication fixes the issue.

cc @linoytsaban @sayakpaul

**Comment by sayakpaul** [comment]
@geniuspatrick

**Comment by bghira** [comment]
it might be a bug in Accelerate, actually. the LR scheduler state isn't being restored correctly when it restarts. I then set it manually, but didn't multiply by num_processes. which led to a mismatch during resume LR. but i think maybe when they are both multiplied correctly. this issue does not manifest. i'm still trying to test it so i can assess the issue on Accelerate's side since I shouldn't / didn't used to have to set the last_epoch manually.

**Comment by sayakpaul** [comment]
Cc: @muellerzr

**Comment by geniuspatrick** [comment]
@AbraarArique In the training scripts, we have two arguments that control the total number of training steps, `args.num_train_epochs` and `args.max_train_steps`(has a higher priority, btw). In your example, you seem to be controlling the number of steps trained through `args.num_train_epochs`. But note that as the number of gpu's changes, so does the number of training steps per epoch.

The argument `args.lr_warmup_steps` would be easier to understand, if together with `args.max_train_steps`. When the number of GPUs changes from 1 to 2, if you keep `args.num_train_epochs=1`, then actually the `args.max_train_steps` has changed from 64 to 32, then you should also adjust the `args.lr_warmup_steps`.

Hopefully the rather redundant explanation above will help your understanding.

**Comment by geniuspatrick** [comment]
> it might be a bug in Accelerate, actually. the LR scheduler state isn't being restored correctly when it restarts. I then set it manually, but didn't multiply by num_processes. which led to a mismatch during resume LR. but i think maybe when they are both multiplied correctly. this issue does not manifest. i'm still trying to test it so i can assess the issue on Accelerate's side since I shouldn't / didn't used to have to set the last_epoch manually.

Looks like a topic about **_resume training_**. Is the scheduler state not being saved and restored correctly? Could the removal of multiplication be just a numerical coincidence? Is it possible to have the same problem with single GPU training?

**Comment by bghira** [comment]
on my end the problem is that we upgraded from accelerate v0.19 to v0.33 and the load/save state for accelerate stopped writing the step count for the random states - or stopped restoring it. either way, i'm on git main now and i have to see if lr_scheduler hasattr last_epoch and set it to resume_step * num_processes.

that fixed my issue for single and multiple GPUs. but i have to leave the num_warmup_steps multiplied too

**Comment by muellerzr** [comment]
Will dig into this @bghira

**Comment by bghira** [comment]
@muellerzr did you ever find anything? it might just be something i've been doing incorrectly but i would like to align with best practices

**Comment by Zephyrose** [comment]
@geniuspatrick @AbraarArique I agree with your point of view, I don't think we should multiply the accelerator.num_processes. Because the lr_warmup_steps/num_training_steps_for_scheduler is a ratio acctually. So no matter how many GPUs we use, we want this ratio to remain constant. When we add GPUs, we are actually increasing the batch size. num_training_steps_for_scheduler is a constant, and actual_training_steps is reduced by num_GPUS times. If we don't multiply the accelerator.num_processes, we will not adjust the args.lr_warmup_steps.

**Comment by AbraarArique** [comment]
@Zephyrose I think there are 2 ways of looking at this:

The way it's done now does make sense logically. If 1 epoch is 32 steps with 1 GPU, with 2 GPUs you're doubling the batch size and thus now you have 16 update steps per epoch.

So the `lr_warmup_steps` being relative to the number of update steps per device/process makes sense: as you add GPUs, you're lowering your update steps and thus lowering your `lr_warmup_steps` accordingly.

This works fine if you specify `max_train_steps`, but as you mentioned, if you're training based on epochs, having to scale the `lr_warmup_steps` down with more GPUs is indeed inconvenient as you often want the ratio to be consistent.

If people care about the warmup-to-total-steps ratio more than a specific number of steps, perhaps it makes sense to have an `lr_warmup_ratio` parameter instead of manually specifying steps?
